### ISSUE TITLE
Create conventions for referencing "unsafe" parameters and their later, safe counterparts

### ISSUE THREAD

**Issue by katelynsills** [Agoric/agoric-sdk#72]
Almost always in defensive programming, all arguments are things we're suspicious of, whether we call this unsafe* or alleged* notReliedOn* or something else. I am certainly in favor of any convention that helps us remember to be suspicious of these! However, it needs to be a convention we can apply consistently. 

With `alleged*` I chose only things for which we must remain suspicious even after the beginning of the function tries to normalize and validate. Thus, `initialBalance` on line 141 is not `allegedInitialBalance` or `unsafeInitialBalance`.

By what principle is `name` worth prefixing but `initialBalance` not?

Regarding declaring new vs reassigning to old, I'm calm about it. We need a convention. But fixing line 142 and not 141 is jarring. ***If*** we don't come up with a naming convention for labeling all parameters we can't rely on, and ***if*** we continue to use unqualified names for most of these, then I would prefer reassignment for normal-named parameter variables that are immediately normalized to relied-upon values on entry to the function, before it does anything else. Otherwise, I agree that declaring new variables is likely better.

We should find principles to create conventions we can widely adhere to.

_Originally posted by @erights in https://github.com/Agoric/ERTP/pull/41_

**Comment by katelynsills** [comment]
I'm going to close this as old and not very actionable. If we want to keep this, let's open a new issue with the particular task to be done. As of 2/8/2021, `assert` is recognized by our jsdoc typescript enforcement.

**Comment by erights** [comment]
Reopening. We need to write down a convention, start following it, and eventually write a static checker for finding code that violates that convention without exempting itself. We have not yet done any of that. Until we do, we should not expect our abstractions to be *defensively consistent*.

**Comment by erights** [comment]
The fact that `assert` propagates type info into the typescript type inference is great! (thx @michaelfig ) It will certainly help this effort enormously. But it is not a substitute for that effort.

**Comment by katelynsills** [comment]
> Reopening. We need to write down a convention, start following it, and eventually write a static checker for finding code that violates that convention without exempting itself. We have not yet done any of that. Until we do, we should not expect our abstractions to be _defensively consistent_.

Ok, sounds good. I'm going to remove the ERTP label then, as this seems more general.

**Comment by erights** [comment]
Yes, this issue is in no way specific to ERTP.

**Comment by erights** [comment]
Hi @dckc I see your sadness above that static analysis tools won't grok this. I'm assigning the two of us in the hope that you'll build static analysis tools that will prove you wrong ;)

**Comment by dckc** [comment]
turns out tsc is smarter than I thought: [asserts condition][1] does what I want. @michaelfig pointed out that we're already using it; for example, [insistCapData][2].

[1]: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#assertion-functions
[2]: https://github.com/Agoric/agoric-sdk/blob/d73f2ce394635cd0244b704250c61eefee4f3819/packages/SwingSet/src/capdata.js#L14

**Comment by dckc** [comment]
@katelynsills this seems relevant to today's discussion of handling arguments defensively.

**Comment by erights** [comment]
Update with some more recent thoughts (from a thread with @dckc ):

@michaelfig is working towards `ExoPromise` which would be a kind of pseudo-promise parameterized by a pattern, guaranteeing that this pseudo-promise can only fulfill to a value that matches that pattern. I expect the pattern to be optional, defaulting to `M.all()`, guaranteeing that an ExoPromise can only fulfill to a `Passable`.

With this, I hope to complete a transition to a hardened JS style where, instead of treating each object boundary by default as a defensive boundary, we can instead treat the remotable call boundary as a defensive boundary, starting with validation by the current Exo interfaces. Many difficulties here yet to be solved. But assuming we succeed...

For the parameters of the raw methods of an exo, the first line of validation defense is the interface checking. For many input validation needs, especially type-like validation needs, this will often be sufficient. When it is not sufficient, and the raw method must still do additional validation of a parameter, this is worth calling attention to. So for exo raw method params, I propose that those params still needing more input validation beyond what the interface already checked, these param names be `alleged*`. 

Essentially, for exos, the 
> after the validations and coercions block at the top of the function

is instead offloaded to the interface checks. The subsequent controversy above about variable reassignment would dissolve away.

We could (and should!) start using this convention of exo raw methods now, without waiting on ExoPromise or the transition to exos as the main (only?) defensiveness boundary.
